diff --git a/dist/clients/consumer/consumer.js b/dist/clients/consumer/consumer.js
--- a/dist/clients/consumer/consumer.js
+++ b/dist/clients/consumer/consumer.js
@@ -567,10 +567,15 @@
                             api(connection, -1, options.isolationLevel ?? this[kOptions].isolationLevel, Array.from(requests.values()), retryCallback);
                         });
                     });
-                }, concurrentCallback, 0);
+                }, concurrentCallback, 0, [], error => error.findBy?.("hasStaleMetadata", true));
             }, (error, responses) => {
                 if (error) {
-                    callback(this.#handleMetadataError(error));
+                    this.#handleMetadataError(error);
+                    if (error.findBy?.("hasStaleMetadata", true)) {
+                        this.#listOffsets(withTimestamps, options, callback);
+                        return;
+                    }
+                    callback(error);
                     return;
                 }
                 let offsets = new Map();
@@ -1007,6 +1012,11 @@
                     this.#performJoinGroup(options, callback);
                     return;
                 }
+                if (error.findBy?.("apiCode", "NOT_COORDINATOR")) {
+                    this.#coordinatorId = undefined;
+                    this.#performJoinGroup(options, callback);
+                    return;
+                }
                 callback(error);
                 return;
             }
@@ -1027,6 +1037,11 @@
                 }
                 if (error) {
                     if (this.#getRejoinError(error)) {
+                        this.#performJoinGroup(options, callback);
+                        return;
+                    }
+                    if (error.findBy?.("apiCode", "NOT_COORDINATOR")) {
+                        this.#coordinatorId = undefined;
                         this.#performJoinGroup(options, callback);
                         return;
                     }
@@ -1183,7 +1198,7 @@
                         }
                         operation(connection, retryCallback);
                     });
-                }, callback);
+                }, callback, 0, [], error => error.findBy?.("apiCode", "NOT_COORDINATOR"));
             });
         });
     }
